shotover-proxy/src/transforms/protect/mod.rs

```diff
@@ -156,25 +156,77 @@ impl ProtectConfig {
     }
 }
 
+pub fn get_values_from_insert_or_update_mut(ast: &mut CQL) -> HashMap<String, &mut SQLValue> {
+    if let CQL::Parsed(ast) = ast {
+        match &mut ast[0] {
+            Statement::Insert {
+                source, columns, ..
+            } => get_values_from_insert_mut(columns.as_mut(), source.borrow_mut()),
+            Statement::Update { assignments, .. } => {
+                get_values_from_update_mut(assignments.as_mut())
+            }
+            _ => HashMap::new(),
+        }
+    } else {
+        HashMap::new()
+    }
+}
+
+fn get_values_from_insert_mut<'a>(
+    columns: &'a mut [Ident],
+    source: &'a mut Query,
+) -> HashMap<String, &'a mut SQLValue> {
+    let mut map = HashMap::new();
+    let mut columns_iter = columns.iter();
+    if let SetExpr::Values(v) = &mut source.body {
+        for value in &mut v.0 {
+            for ex in value {
+                if let Expr::Value(v) = ex {
+                    if let Some(c) = columns_iter.next() {
+                        map.insert(c.value.to_string(), v);
+                    }
+                }
+            }
+        }
+    }
+    map
+}
+
+fn get_values_from_update_mut(assignments: &mut [Assignment]) -> HashMap<String, &mut SQLValue> {
+    let mut map = HashMap::new();
+    for assignment in assignments {
+        if let Expr::Value(v) = &mut assignment.value {
+            map.insert(assignment.id.iter().map(|x| &x.value).join("."), v);
+        }
+    }
+    map
+}
+
 #[async_trait]
 impl Transform for Protect {
     async fn transform<'a>(&'a mut self, mut message_wrapper: Wrapper<'a>) -> ChainResponse {
+        // encrypt the values included in any INSERT or UPDATE queries
         for message in message_wrapper.messages.iter_mut() {
-            if let MessageDetails::Query(qm) = &mut message.details {
-                // Encrypt the writes
-                if QueryType::Write == qm.query_type && qm.namespace.len() == 2 {
-                    if let Some((_, tables)) =
-                        self.keyspace_table_columns.get_key_value(&qm.namespace[0])
+            if let Some(namespace) = message.namespace() {
+                if namespace.len() == 2 {
```

Comment:
Won't a namespace always have the keyspace and column? Once we fix the `USE keyspace` handling of course.

Reply:
I'm just maintaining the existing functionality, I think when we implement proper `USE keyspace` we can go through and make sure the namespace implementation and usage is sensible.